### PR TITLE
fix: recreate watchers after nats connection reset

### LINKED ISSUE #864: OT store watcher stopped working
**Describe the bug**

OT store watcher stopped working for some reason (meanwhile PROCESSOR watcher still works, and the data processing is still going on), this made the OT in the memory stop updating, and the watermark also stopped propagation.

Do not see any JetStream pod restart, also no reconnection in the pod log.

See the logs attached.

[logs.txt](https://github.com/numaproj/numaflow/files/12059041/logs.txt)

**Screenshots**

**Environment (please complete the following information):**

- Kubernetes: [e.g. v1.18.6]
- Numaflow: [e.g. v0.9.0-rc1

**Additional context**

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by yhl25**
if the nats connection is reset we need to recreate the watchers, otherwise the watcher stops getting updates.

fixes #864

### CODE REVIEW

File: pkg/watermark/store/jetstream/kv_watch.go [hunk 1]
```diff
@@ -35,45 +34,62 @@ type jetStreamWatch struct {
...
+		kvBucketName: kvBucketName,
+		kvWatchers:   make([]*KeyWatcher, 0),
+		log:          logging.FromContext(ctx).With("pipeline", pipelineName).With("kvBucketName", kvBucketName),
 	}
 
-	js, err := conn.JetStream()
-	if err != nil {
-		if !conn.IsClosed() {
+	connectAndWatch := func() (*jsclient.NatsConn, *jsclient.JetStreamContext, error) {
+		conn, err := client.Connect(ctx, jsclient.ReconnectHandler(func(c *jsclient.NatsConn) {
```
Comment (whynowy) [inline pkg/watermark/store/jetstream/kv_watch.go:53]: Can this also handle the nats native auto reconnect caused problem?

Comment (yhl25) [reply pkg/watermark/store/jetstream/kv_watch.go:53]: After several attempts, I noticed watcher stops working with nats native auto-reconnect as well and therefore modified the logic to disable it.

**Comment by vigith** [comment]
don't we need locks to update connections?

File: pkg/shared/clients/nats/default_jetstream_client.go [hunk 2]
```diff
@@ -63,9 +63,6 @@ func natsJetStreamConnection(ctx context.Context, url string, natsOptions []nats
 				log.Info("Nats: disconnected")
 			}
 		}),
-		nats.ReconnectHandler(func(nnc *nats.Conn) {
```
Comment (whynowy) [inline pkg/shared/clients/nats/default_jetstream_client.go:66]: No harm to keep it?

Comment (whynowy) [reply pkg/shared/clients/nats/default_jetstream_client.go:66]: BTW - does like 58 need to be changed?

Comment (yhl25) [reply pkg/shared/clients/nats/default_jetstream_client.go:66]: I am defining reconnect handler `inside in_cluster_jetstream_client.go`

**Comment by yhl25** [comment]
closing this since this is handled in #889

**Pull request closed without merging**